[DAG] Generalize setcc(setcc) fold to use known bits.

davemgreen opened the pull request:
If we have a `SETCC (SETCC), 0, NE` and ZeroOrOneBooleanContent, we can remove
the outer setcc as it will produce the same value as the inner. This can be
generalized to anything where the top bits are known to be 0, as the value will
remain as 1 or 0.

davemgreen committed:
[AArch64] Add some tests for setcc known bits fold. NFC

goldsteinn commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -4652,21 +4652,25 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
-      // SETCC (SETCC), [0|1], [EQ|NE]  -> SETCC
-      if (N0.getOpcode() == ISD::SETCC &&
+      // SETCC (X), [0|1], [EQ|NE]  -> X if X is known 0/1
+      if ((N0.getOpcode() == ISD::SETCC || VT != MVT::i1) &&
```
I don't quite understand the `VT != MVT::i1` here. Doesn't that break things like `setcc (setcc ...))`

davemgreen replied:
It will do the transform if it is a setcc, or if the type _isn't_ an i1. i1 types are trivially true for the known bits combine, and handled further below when the foldBools flag is set.

davemgreen committed:
[DAG] Generalize setcc(setcc) fold to use known bits.

If we have a `SETCC (SETCC), 0, NE` and ZeroOrOneBooleanContent, we can remove
the outer setcc as it will produce the same value as the inner. This can be
generalized to anything where the top bits are known to be 0, as the value will
remain as 1 or 0.

RKSimon commented on llvm/test/CodeGen/X86/xor.ll:
```diff
@@ -403,14 +403,19 @@ define i32 @PR17487(i1 %tobool) {
+; X64-WIN-NEXT:    movd %eax, %xmm0
+; X64-WIN-NEXT:    pshufd {{.*#+}} xmm0 = xmm0[0,1,0,1]
+; X64-WIN-NEXT:    pand __xmm@00000000000000010000000000000001(%rip), %xmm0
+; X64-WIN-NEXT:    pextrw $4, %xmm0, %eax
```
Interfering with scalarizeExtractedBinop ?

davemgreen replied:
Yeah these I wasn't sure about - how important they were. They get folded away under normal optimizations: https://godbolt.org/z/P4vMYd5oc, and it looked like the test was only added for correctness.

I think if the combines happened in topo order it might do better. The DAG looks like this:
```
          t56: v4i32 = BUILD_VECTOR undef:i32, undef:i32, t2, undef:i32
        t57: v2i64 = bitcast t56                                       
        t36: v2i64 = BUILD_VECTOR Constant:i64<1>, Constant:i64<1>     
      t39: v2i64 = and t57, t36                                        
    t50: v16i8 = bitcast t39                                           
  t51: i8 = extract_vector_elt t50, Constant:i64<8>                    
t45: i32 = zero_extend t51                                             
```

goldsteinn replied:
I'm not really an authority on this, but the combine code makes sense so I wouldn't think its a dealbreaker, but would nice if we could fix it up.

RKSimon replied:
So my hunch would be that scalarizeExtractedBinop can be extended to permit the fold:
```
extract_vector_elt(oneusebitcast(bitlogicop(x,y)))
 ->
bitlogicop(extract_vector_elt(bitcast(x)),extract_vector_elt(bitcast(y)))
```

goldsteinn replied:
Hmm, would imagine the 2x `extract_vector_elt` would be slower. Why is that advantagious?

RKSimon replied:
scalarizeExtractedBinop only performs this if at least one of the operands can freely extract (i.e. its a constant)

RKSimon replied:
@davemgreen Any luck with this?

arsenm reviewed (approved):
lgtm. The small test regression is unrelated

RKSimon reviewed (approved):
OK, no objection from me

arsenm committed:
Merge branch 'main' into gh-dag-setccknownbits
